docs/rewriter/design.md

```diff
@@ -0,0 +1,130 @@
+The current implementation supports one version of such a language.
+
+Option 1 has the advantage that we reuse the existing implementation,
+and users already familiar with ONNX Script can directly use it.
+However, this approach has some limitations. For example, in the
+pattern we may not care about a required-attribute's value. If
+ONNX Script complains about a missing required-attribute, that would
+be a problem when specifying a pattern.
+
+Option 2 allows us to have a more extensible mechanism, customized to
+the case of patterns. For example, it is easy to allow users to specify
+that we are looking for attribute `alpha` to have a value of `0.16325`
+within some error tolerance limits. (Note that we can push such conditions
+into the guard/condition function too. So, this specific example is more
+of a syntactic-sugar convenience.) Another interesting use-case, which
+comes up in practice, is when we want to do some wild-card matching against
+a op/function name ... eg., if we want to match against a call to a
+function (produced by the pytorch exporter) with a name that has `Silu` as
+a prefix.
+
+## IR and Builder for Patterns and Rewrite-Rules
+
+We already have a couple of approaches to pattern-matching based rewrite-rules:
+the original one (restricted to single-output patterns, but supporting other
+features such as a mechanism to bind variables to specific nodes of the
+pattern, and using them to check for extra-conditions before applying a
+rewrite-rule, and allowing for more efficient match-and-replace), and a
+newer extension to support multiple-output patterns, but not integrated
+with the other mechanisms.
+
+We can integrate and simplify these approaches, including the different options
+for describing patterns mentioned above, by unifying the classes defining
+the patterns and rewrite-rules (into a pattern IR) and by providing different
+factory (constructor) methods for patterns and rewrite-rules. Specifically,
+these methods will be the developer-facing APIs for expressing rewrite-rules.
+Hence, it is important for us to consolidate these APIs to avoid potential
+confusion among users writing rewrite-rules. (The implementations of the
+pattern-matching can vary, depending on the complexity of the pattern,
+such as whether it is single-output or multiple-output, etc.)
+
+Proposal:
+
+* The rewrite-rule constructor takes two parameters, one to specify the
+pattern, and one to specify the replacement. We can provide a version with
+an extra condition parameter (as a convenience feature).
+
+* The pattern can be either a FunctionProto (or, equivalently, an ONNX Script
+function) or a function that constructs a pattern-graph or a pattern-graph.
+In the first two cases, they are converted into a pattern-graph.
+
+* The replacement is a function that accepts an expanded match-bindings
+(that is, `**match_bindings`) and returns either `None` (in case of failure)
```

Comment:
Is this for the first part (of searching for a pattern-match), which is Xavier's point below? The failure mentioned in the above line is in user's code (the replacement-function), so it is presumably easier for the user to track this (they can return an appropriate error message they choose instead of just "None", I guess).

For the first part, see discussion below.